lib/inferno/dsl/fhir_evaluation/rules/all_resources_reachable.rb

```diff
@@ -0,0 +1,76 @@
+# frozen_string_literal: true
+
+require_relative '../../../utils/evaluator_util'
+
+module Inferno
+  module DSL
+    module FHIREvaluation
+      module Rules
+        class AllResourcesReachable < Rule
+          attr_accessor :config
+
+          def check(context)
+            @config = context.config
+            # TODO: add some customizable configurations for this rule:
+            # - whether it should be checked at all
+            # - determine the base/root resource
+            # - which types/resources to ignore
+            # TODO: for the customizable configurations, use the config settings below:
+            # config.Rule.AllResourcesReachable.CheckedAtAll: true/false
+            # config.Rule.AllResourcesReachable.RunOnlyBaseResource: true/false
+            # config.Rule.AllResourcesReachable.IgnoreType: - value
+            # TODO: can come up with a "connectedness metric" to see how well-connected the data set is
+
+            # every resource is either making a resolvable reference or is referenced
+            util = Inferno::Utils::EvaluatorUtil
+            @referenced_resources = Set.new
+            @referencing_resources = Set.new
+            @resourcetype_ids, @resource_type_ids, @resource_ids, references = util.extract_ids_references(context.data)
+            references.each do |id, refs|
+              assess_reachability(id, refs)
+            end
+
+            island_resources = @resource_ids - @referenced_resources - @referencing_resources
+            sorted_island_resources = []
+            island_resources.each do |id|
+              types = @resourcetype_ids.select { |_key, values| values.include?(id) }.keys
+              sorted_island_resources << ("#{types[0]}/#{id}")
+            end
+
+            if sorted_island_resources.any?
+              message = "Found resources that have no resolved references and are not referenced: #{
+            sorted_island_resources.join(', ')}"
+              result = EvaluationResult.new(message, rule: self)
+            else
+              message = 'All resources are reachable'
+              result = EvaluationResult.new(message, severity: 'success', rule: self)
+            end
+
+            context.add_result result
```

Comment:
I feel like this should return results rather than mutating the context.

Reply:
Addressed

Reply:
Dylan's comments for this:
https://github.com/inferno-framework/inferno-core/pull/583#discussion_r1905935612